jib-plugins-common/src/main/java/com/google/cloud/tools/jib/plugins/common/JavaContainerBuilderHelper.java

```diff
@@ -64,33 +69,40 @@ public static FileEntriesLayer extraDirectoryLayerConfiguration(
           FileSystems.getDefault().getPathMatcher("glob:" + entry.getKey()), entry.getValue());
     }
 
-    new DirectoryWalker(sourceDirectory)
-        .filterRoot()
-        .walk(
-            localPath -> {
-              AbsoluteUnixPath pathOnContainer =
-                  targetDirectory.resolve(sourceDirectory.relativize(localPath));
-              Instant modificationTime = modificationTimeProvider.get(localPath, pathOnContainer);
-              FilePermissions permissions =
-                  extraDirectoryPermissions.get(pathOnContainer.toString());
-              if (permissions == null) {
-                // Check for matching globs
-                Path containerPath = Paths.get(pathOnContainer.toString());
-                for (Map.Entry<PathMatcher, FilePermissions> entry : pathMatchers.entrySet()) {
-                  if (entry.getKey().matches(containerPath)) {
-                    builder.addEntry(
-                        localPath, pathOnContainer, entry.getValue(), modificationTime);
-                    return;
-                  }
-                }
-
-                // Add with default permissions
-                builder.addEntry(localPath, pathOnContainer, modificationTime);
-              } else {
-                // Add with explicit permissions
-                builder.addEntry(localPath, pathOnContainer, permissions, modificationTime);
+    DirectoryWalker walker = new DirectoryWalker(sourceDirectory).filterRoot();
+    excludes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .forEach(pathMatcher -> walker.filter(path -> !pathMatcher.matches(path)));
+    includes
+        .stream()
+        .map(pattern -> FileSystems.getDefault().getPathMatcher("glob:" + pattern))
+        .map(pathMatcher -> (Predicate<Path>) (path -> pathMatcher.matches(path)))
+        .reduce((matches1, matches2) -> matches1.or(matches2))
+        .ifPresent(walker::filter);
+    walker.walk(
+        localPath -> {
+          AbsoluteUnixPath pathOnContainer =
+              targetDirectory.resolve(sourceDirectory.relativize(localPath));
+          Instant modificationTime = modificationTimeProvider.get(localPath, pathOnContainer);
+          FilePermissions permissions = extraDirectoryPermissions.get(pathOnContainer.toString());
+          if (permissions == null) {
+            // Check for matching globs
+            Path containerPath = Paths.get(pathOnContainer.toString());
+            for (Map.Entry<PathMatcher, FilePermissions> entry : pathMatchers.entrySet()) {
+              if (entry.getKey().matches(containerPath)) {
+                builder.addEntry(localPath, pathOnContainer, entry.getValue(), modificationTime);
+                return;
```

Comment:
there seem to be too many levels of nesting here, can we put this in a helper method?